File: SimplePhysics/PhysicsBody.cs
using System;
using System.Collections.Generic;
using System.Text;
using Jypeli;

namespace Jypeli.Physics
{
    [Save]
    public class PhysicsBody : IPhysicsBody
    {
        [Save] private double _mass = 1.0;
        [Save] private double _massInv = 1.0;
        [Save] private double _restitution = 1.0;
        [Save] private double _staticFriction = 0.0;
        [Save] private double _kineticFriction = 0.0;

        public event CollisionHandler<IPhysicsBody, IPhysicsBody> Collided;

        [Save]
        public IPhysicsObject Owner { get; internal set; }

        [Save]
        public Shape Shape { get; private set; }

        public double Mass
        {
            get { return _mass; }
            set
            {
                _mass = value;
                _massInv = 1 / value;
            }
        }

        public double MassInv
        {
            get { return _massInv; }
            set
            {
                _massInv = value;
                _mass = 1 / value;
            }
        }

        public double MomentOfInertia
        {
            get { return double.PositiveInfinity; }
            set { }
        }

        public double Restitution
        {
            get { return _restitution; }
            set { _restitution = value; }
        }

        public double KineticFriction
        {
            get { return _kineticFriction; }
            set { _kineticFriction = value; }
        }

        public double StaticFriction
        {
            get { return _staticFriction; }
            set { _staticFriction = value; }
        }

        public double AngularDamping
        {
            get { return 0; }
            set { }
        }

        public double LinearDamping
        {
            get { return 0; }
            set { }
        }

        [Save]
        public Vector Position { get; set; }

        [Save]
        public Vector Velocity { get; set; }

        [Save]
        public Vector Acceleration { get; set; }

        [Save]
        public Vector Size { get; set; }

        public double Angle
        {
            get { return 0; }
            set { }
        }

        public double AngularVelocity
        {
            get { return 0; }
            set { }
        }

        public double AngularAcceleration
        {
            get { return 0; }
            set { }
        }

        public bool IgnoresCollisionResponse { get; set; }

        public bool IgnoresGravity { get; set; }

        public bool IgnoresPhysicsLogics { get; set; }

        public PhysicsBody( double width, double height, Shape shape )
        {
            Shape = shape;
            Size = new Vector( width, height );
        }

        public void MakeStatic()
        {
            _mass = double.PositiveInfinity;
            _massInv = 0;
        }

        public void ApplyForce( Vector force )
        {
            Acceleration += force * _massInv;
        }

        public void ApplyImpulse( Vector impulse )
        {
            Velocity += impulse * _massInv;
        }

        public void ApplyTorque( double torque )
        {
            // No angular motion implemented (yet?)
        }

        public void Stop()
        {
            Acceleration = Vector.Zero;
            Velocity = Vector.Zero;
        }

        public void StopAngular()
        {
            // No angular motion to stop :)
        }

        public void StopAxial( Vector axis )
        {
            Acceleration = Acceleration.Project( axis.LeftNormal );
            Velocity = Velocity.Project( axis.LeftNormal );
        }
    }
}

File: SimplePhysics/SimplePhysics-FS2010/PhysicsBody.fs
// Learn more about F# at http://fsharp.net

module SimplePhysics

open Jypeli
open Jypeli.Physics

type public PhysicsBody(w, h, s) =
    let mutable _shape : Shape = s
    let mutable _mass  = 1.0
    let mutable _massInv = 1.0;
    let mutable _restitution = 1.0

    let mutable _size = new Vector(w, h)
    let mutable _pos = Vector.Zero
    let mutable _vel = Vector.Zero
    let mutable _acc = Vector.Zero

    let mutable _iCollResp = false
    let mutable _iGrav = false
    let mutable _iPhysLog = false

    member this.Shape
        with set(a) = _shape <- a

    member this.Size
        with get() = _size
        and set(a) = _size <- a

    interface IPhysicsBody with
        member this.Shape with get() = _shape
        member this.Mass
            with get() = _mass
            and set(a) =
                _mass <- a
                _massInv <- 1.0 / a
        member this.MassInv
            with get() = _massInv
            and set(a) =
                _massInv <- a
                _mass <- 1.0 / a
        member this.Restitution
            with get() = _restitution
            and set(a) = _restitution <- a

        member this.MomentOfInertia
            with get() = infinity
            and set(a) = ()
        member this.KineticFriction
            with get() = 0.0
            and set(a) = ()
        member this.AngularDamping
            with get() = 1.0
            and set(a) = ()
        member this.LinearDamping
            with get() = 1.0
            and set(a) = ()

        member this.Position
            with get() = _pos
            and set(a) = _pos <- a
        member this.Velocity
            with get() = _vel
            and set(a) = _vel <- a
        member this.Acceleration
            with get() = _acc
            and set(a) = _acc <- a

        member this.Angle
            with get() = 0.0
            and set(a) = ()
        member this.AngularVelocity
            with get() = 0.0
            and set(a) = ()
        member this.AngularAcceleration
            with get() = 0.0
            and set(a) = ()

        member this.IgnoresCollisionResponse
            with get() = _iCollResp
            and set(a) = _iCollResp <- a
        member this.IgnoresGravity
            with get() = _iGrav
            and set(a) = _iGrav <- a
        member this.IgnoresPhysicsLogics
            with get() = _iPhysLog
            and set(a) = _iPhysLog <- a

        member this.MakeStatic() =
            _mass <- infinity
            _massInv <- 0.0
            _iGrav <- true

        member this.ApplyImpulse(impulse) =
            _vel <- _vel + impulse * _massInv

